File: src/types/template.ts
// === Jasper Template TypeScript Types ===
// This file matches the Rust structures for type safety

// === Template Metadata ===
export interface TemplateMetadata {
  readonly version: string
  readonly format_version: string
  readonly created_at: string // ISO 8601 format
  readonly last_modified: string // ISO 8601 format
  readonly created_by: string
  readonly description?: string
  readonly tags: ReadonlyArray<string>
  readonly compatibility: CompatibilityInfo
}

export interface CompatibilityInfo {
  readonly min_jasper_version: string
  readonly jasperreports_version?: string
}

// === Canvas Configuration ===
export interface Canvas {
  readonly width: number
  readonly height: number
  readonly unit: PageUnit
  readonly orientation: PageOrientation
  readonly margins: PageMargins
  readonly grid: GridConfig
  readonly background: BackgroundConfig
}

export type PageUnit = 'pt' | 'mm' | 'in' | 'px'
export type PageOrientation = 'portrait' | 'landscape'

export interface PageMargins {
  readonly top: number
  readonly bottom: number
  readonly left: number
  readonly right: number
}

export interface GridConfig {
  readonly enabled: boolean
  readonly size: number
  readonly snap: boolean
  readonly visible: boolean
}

export interface BackgroundConfig {
  readonly color: string
  readonly image?: string
}

// === Data Source Configuration ===
export interface DataSource {
  readonly id: string
  readonly name: string
  readonly source_type: DataSourceType
  readonly provider_type: string
  readonly config: DataSourceConfig
  readonly schema: DataSchema
  readonly query?: DataQuery
}

export type DataSourceType = 'json' | 'excel' | 'sql' | 'xml' | 'csv'

export interface DataSourceConfig {
  // Database configuration
  readonly host?: string
  readonly port?: number
  readonly database?: string
  readonly username?: string
  readonly password?: string
  readonly ssl_mode?: string
  
  // File configuration
  readonly file_path?: string
  readonly sheet_name?: string
  readonly delimiter?: string
  
  // API configuration
  readonly url?: string
  readonly headers?: Record<string, string>
  readonly auth_method?: string
  
  // Connection settings
  readonly timeout?: number
  readonly retry_count?: number
}

export interface DataSchema {
  readonly columns: ReadonlyArray<ColumnInfo>
  readonly primary_key?: string
  readonly indexes: ReadonlyArray<string>
  readonly relationships: Record<string, any>
  readonly metadata: Record<string, any>
}

export interface ColumnInfo {
  readonly name: string
  readonly data_type: DataType
  readonly nullable: boolean
  readonly default_value?: any
  readonly constraints: ReadonlyArray<string>
  readonly description?: string
  readonly format_hint?: string
}

export type DataType = 'String' | 'Number' | 'Boolean' | 'Date' | 'DateTime' | 'Array' | 'Object' | 'Null'

export interface DataQuery {
  readonly sql?: string
  readonly path?: string
  readonly filter?: string
  readonly limit?: number
  readonly offset?: number
  readonly sort?: string
  readonly parameters: ReadonlyArray<QueryParameter>
}

export interface QueryParameter {
  readonly name: string
  readonly param_type: DataType
  readonly default?: any
  readonly description?: string
}

// === Template Elements ===
export interface TemplateElement {
  readonly id: string
  readonly element_type: ElementType
  readonly position: Position
  readonly size: Size
  readonly z_index: number
  readonly visible: boolean
  readonly content: ElementContent
  readonly style: ElementStyle
  readonly data_binding?: DataBinding
}

export type ElementType = 'Text' | 'DataField' | 'Rectangle' | 'Line' | 'Image' | 'Barcode' | 'Chart'

export interface Position {
  readonly x: number
  readonly y: number
}

export interface Size {
  readonly width: number
  readonly height: number
}

export interface ElementContent {
  readonly text?: string
  readonly expression?: string
  readonly font?: FontConfig
  readonly alignment?: AlignmentConfig
  readonly color?: string
  readonly format?: string
}

export interface FontConfig {
  readonly family: string
  readonly size: number
  readonly weight: FontWeight
  readonly style: FontStyle
}

export type FontWeight = 'normal' | 'bold' | 'light'
export type FontStyle = 'normal' | 'italic' | 'oblique'

export interface AlignmentConfig {
  readonly horizontal: HorizontalAlignment
  readonly vertical: VerticalAlignment
}

export type HorizontalAlignment = 'left' | 'center' | 'right' | 'justify'
export type VerticalAlignment = 'top' | 'middle' | 'bottom'

export interface ElementStyle {
  readonly background?: BackgroundStyle
  readonly border?: BorderStyle
  readonly padding?: PaddingConfig
}

export interface BackgroundStyle {
  readonly color?: string
  readonly image?: string
}

export interface BorderStyle {
  readonly width: number
  readonly color: string
  readonly style: BorderStyleType
}

export type BorderStyleType = 'solid' | 'dashed' | 'dotted'

export interface PaddingConfig {
  readonly top: number
  readonly bottom: number
  readonly left: number
  readonly right: number
}

export interface DataBinding {
  readonly source_id: string
  readonly field_name: string
  readonly validation?: ValidationRules
}

export interface ValidationRules {
  readonly required: boolean
  readonly max_length?: number
  readonly min_length?: number
  readonly pattern?: string
}

// === Parameters and Variables ===
export interface Parameter {
  readonly name: string
  readonly param_type: DataType
  readonly default?: any
  readonly description?: string
  readonly required: boolean
}

export interface Variable {
  readonly name: string
  readonly var_type: DataType
  readonly expression: string
  readonly initial_value?: any
  readonly description?: string
}

// === Groups ===
export interface Group {
  readonly name: string
  readonly expression: string
  readonly sort_order: SortOrder
  readonly header?: BandConfig
  readonly footer?: BandConfig
}

export type SortOrder = 'asc' | 'desc'

export interface BandConfig {
  readonly height: number
  readonly elements: ReadonlyArray<string> // Element IDs
}

// === Main Template Structure ===
export interface JasperTemplate {
  readonly metadata: TemplateMetadata
  readonly canvas: Canvas
  readonly data_sources: ReadonlyArray<DataSource>
  readonly elements: ReadonlyArray<TemplateElement>
  readonly parameters: ReadonlyArray<Parameter>
  readonly variables: ReadonlyArray<Variable>
  readonly groups: ReadonlyArray<Group>
}

// === Template Information and Utilities ===
export interface TemplateInfo {
  readonly version: string
  readonly format_version: string
  readonly created_at: string
  readonly last_modified: string
  readonly description?: string
  readonly tags: ReadonlyArray<string>
  readonly element_count: number
  readonly data_source_count: number
  readonly parameter_count: number
}

export interface TemplatePreview {
  readonly metadata: TemplateMetadata
  readonly canvas: CanvasInfo
  readonly elements: ReadonlyArray<ElementSummary>
  readonly data_sources: ReadonlyArray<DataSourceSummary>
  readonly parameters: ReadonlyArray<Parameter>
  readonly variables: ReadonlyArray<Variable>
}

export interface CanvasInfo {
  readonly width: number
  readonly height: number
  readonly unit: string
  readonly orientation: string
}

export interface ElementSummary {
  readonly id: string
  readonly element_type: string
  readonly position: Position
  readonly size: Size
  readonly visible: boolean
}

export interface DataSourceSummary {
  readonly id: string
  readonly name: string
  readonly source_type: string
  readonly column_count: number
}

// === Template Format Types ===
export type TemplateFormat = 'json' | 'binary' | 'jrxml' | 'unknown'

// === Validation Results ===
export interface TemplateValidationResult {
  readonly valid: boolean
  readonly errors: ReadonlyArray<string>
  readonly warnings: ReadonlyArray<string>
}

// === Factory Functions for Creating Templates ===
export const createEmptyTemplate = (): Partial<JasperTemplate> => ({
  metadata: {
    version: '2.0.0',
    format_version: '1.0.0',
    created_at: new Date().toISOString(),
    last_modified: new Date().toISOString(),
    created_by: 'jasper-designer-v2',
    tags: [],
    compatibility: {
      min_jasper_version: '2.0.0',
      jasperreports_version: '6.20.0',
    },
  },
  canvas: {
    width: 595.0,
    height: 842.0,
    unit: 'pt' as PageUnit,
    orientation: 'portrait' as PageOrientation,
    margins: {
      top: 20.0,
      bottom: 20.0,
      left: 20.0,
      right: 20.0,
    },
    grid: {
      enabled: true,
      size: 10.0,
      snap: true,
      visible: true,
    },
    background: {
      color: '#ffffff',
    },
  },
  data_sources: [],
  elements: [],
  parameters: [],
  variables: [],
  groups: [],
})

export const createDefaultFont = (): FontConfig => ({
  family: 'Arial',
  size: 12.0,
  weight: 'normal',
  style: 'normal',
})

export const createDefaultAlignment = (): AlignmentConfig => ({
  horizontal: 'left',
  vertical: 'middle',
})

export const createDefaultPadding = (): PaddingConfig => ({
  top: 0.0,
  bottom: 0.0,
  left: 0.0,
  right: 0.0,
})

// === Element Creation Helpers ===
export const createTextElement = (
  text: string,
  position: Position,
  size: Size,
  font?: Partial<FontConfig>
): Partial<TemplateElement> => ({
  id: crypto.randomUUID(),
  element_type: 'Text',
  position,
  size,
  z_index: 1,
  visible: true,
  content: {
    text,
    font: { ...createDefaultFont(), ...font },
    alignment: createDefaultAlignment(),
    color: '#000000',
  },
  style: {
    padding: createDefaultPadding(),
  },
})

export const createDataFieldElement = (
  expression: string,
  position: Position,
  size: Size,
  dataBinding?: DataBinding
): Partial<TemplateElement> => ({
  id: crypto.randomUUID(),
  element_type: 'DataField',
  position,
  size,
  z_index: 1,
  visible: true,
  content: {
    expression,
    font: createDefaultFont(),
    alignment: createDefaultAlignment(),
    color: '#000000',
  },
  style: {
    padding: createDefaultPadding(),
  },
  ...(dataBinding && { data_binding: dataBinding }),
})

export const createRectangleElement = (
  position: Position,
  size: Size,
  fillColor?: string,
  borderStyle?: BorderStyle
): Partial<TemplateElement> => ({
  id: crypto.randomUUID(),
  element_type: 'Rectangle',
  position,
  size,
  z_index: 0,
  visible: true,
  content: {
    ...(fillColor && { color: fillColor }),
  },
  style: {
    ...(fillColor && { background: { color: fillColor } }),
    ...(borderStyle && { border: borderStyle }),
    padding: createDefaultPadding(),
  },
})

// === Type Guards ===
export const isTextElement = (element: TemplateElement): boolean => 
  element.element_type === 'Text'

export const isDataFieldElement = (element: TemplateElement): boolean => 
  element.element_type === 'DataField'

export const isRectangleElement = (element: TemplateElement): boolean => 
  element.element_type === 'Rectangle'

export const isLineElement = (element: TemplateElement): boolean => 
  element.element_type === 'Line'

export const isImageElement = (element: TemplateElement): boolean => 
  element.element_type === 'Image'

export const hasDataBinding = (element: TemplateElement): boolean => 
  element.data_binding !== undefined

// === Template Utilities ===
export const getElementsInBounds = (
  template: JasperTemplate,
  bounds: { x: number; y: number; width: number; height: number }
): ReadonlyArray<TemplateElement> => {
  return template.elements.filter(element => {
    const elementRight = element.position.x + element.size.width
    const elementBottom = element.position.y + element.size.height
    const boundsRight = bounds.x + bounds.width
    const boundsBottom = bounds.y + bounds.height
    
    return !(
      element.position.x > boundsRight ||
      elementRight < bounds.x ||
      element.position.y > boundsBottom ||
      elementBottom < bounds.y
    )
  })
}

export const getElementById = (
  template: JasperTemplate,
  id: string
): TemplateElement | undefined => {
  return template.elements.find(element => element.id === id)
}

export const getDataSourceById = (
  template: JasperTemplate,
  id: string
): DataSource | undefined => {
  return template.data_sources.find(ds => ds.id === id)
}

export const validateExpression = (
  expression: string,
  template: JasperTemplate
): { valid: boolean; error?: string } => {
  // Basic expression validation
  if (!expression.trim()) {
    return { valid: false, error: 'Expression cannot be empty' }
  }
  
  // Check if expression uses valid syntax (basic check)
  const fieldPattern = /\{([^}]+)\}/g
  const matches = [...expression.matchAll(fieldPattern)]
  
  for (const match of matches) {
    const fieldPath = match[1]
    if (!fieldPath) continue
    
    // Validate field exists in data sources
    let fieldFound = false
    
    for (const dataSource of template.data_sources) {
      const topLevelField = fieldPath.split('.')[0]
      if (dataSource.schema.columns.some((col: any) => col.name === topLevelField)) {
        fieldFound = true
        break
      }
    }
    
    if (!fieldFound) {
      return { valid: false, error: `Field '${fieldPath}' not found in any data source` }
    }
  }
  
  return { valid: true }
}

File: src-tauri/src/core/template.rs
// === Jasper Template Serialization Core Types ===
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use crate::core::element::{Position, Size, ReportElement};
use crate::errors::{AppError, Result};
use chrono::{DateTime, Utc};

// === Template Metadata ===
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TemplateMetadata {
    pub version: String,
    pub format_version: String,
    pub created_at: DateTime<Utc>,
    pub last_modified: DateTime<Utc>,
    pub created_by: String,
    pub description: Option<String>,
    pub tags: Vec<String>,
    pub compatibility: CompatibilityInfo,
}

impl Default for TemplateMetadata {
    fn default() -> Self {
        let now = Utc::now();
        Self {
            version: "2.0.0".to_string(),
            format_version: "1.0.0".to_string(),
            created_at: now,
            last_modified: now,
            created_by: "jasper-designer-v2".to_string(),
            description: None,
            tags: Vec::new(),
            compatibility: CompatibilityInfo::default(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CompatibilityInfo {
    pub min_jasper_version: String,
    pub jasperreports_version: Option<String>,
}

impl Default for CompatibilityInfo {
    fn default() -> Self {
        Self {
            min_jasper_version: "2.0.0".to_string(),
            jasperreports_version: Some("6.20.0".to_string()),
        }
    }
}

// === Canvas Configuration ===
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Canvas {
    pub width: f64,
    pub height: f64,
    pub unit: PageUnit,
    pub orientation: PageOrientation,
    pub margins: PageMargins,
    pub grid: GridConfig,
    pub background: BackgroundConfig,
}

impl Default for Canvas {
    fn default() -> Self {
        Self {
            width: 595.0, // A4 width in points
            height: 842.0, // A4 height in points
            unit: PageUnit::Point,
            orientation: PageOrientation::Portrait,
            margins: PageMargins::default(),
            grid: GridConfig::default(),
            background: BackgroundConfig::default(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum PageUnit {
    #[serde(rename = "pt")]
    Point,
    #[serde(rename = "mm")]
    Millimeter,
    #[serde(rename = "in")]
    Inch,
    #[serde(rename = "px")]
    Pixel,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum PageOrientation {
    #[serde(rename = "portrait")]
    Portrait,
    #[serde(rename = "landscape")]
    Landscape,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PageMargins {
    pub top: f64,
    pub bottom: f64,
    pub left: f64,
    pub right: f64,
}

impl Default for PageMargins {
    fn default() -> Self {
        Self {
            top: 20.0,
            bottom: 20.0,
            left: 20.0,
            right: 20.0,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GridConfig {
    pub enabled: bool,
    pub size: f64,
    pub snap: bool,
    pub visible: bool,
}

impl Default for GridConfig {
    fn default() -> Self {
        Self {
            enabled: true,
            size: 10.0,
            snap: true,
            visible: true,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BackgroundConfig {
    pub color: String,
    pub image: Option<String>,
}

impl Default for BackgroundConfig {
    fn default() -> Self {
        Self {
            color: "#ffffff".to_string(),
            image: None,
        }
    }
}

// === Data Source Configuration ===
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DataSource {
    pub id: String,
    pub name: String,
    pub source_type: DataSourceType,
    pub provider_type: String,
    pub config: DataSourceConfig,
    pub schema: DataSchema,
    pub query: Option<DataQuery>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum DataSourceType {
    Json,
    Excel,
    Sql,
    Xml,
    Csv,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DataSourceConfig {
    // Database configuration
    pub host: Option<String>,
    pub port: Option<u16>,
    pub database: Option<String>,
    pub username: Option<String>,
    pub password: Option<String>,
    pub ssl_mode: Option<String>,
    
    // File configuration
    pub file_path: Option<String>,
    pub sheet_name: Option<String>,
    pub delimiter: Option<String>,
    
    // API configuration
    pub url: Option<String>,
    pub headers: Option<HashMap<String, String>>,
    pub auth_method: Option<String>,
    
    // Connection settings
    pub timeout: Option<u64>,
    pub retry_count: Option<u32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DataSchema {
    pub columns: Vec<ColumnInfo>,
    pub primary_key: Option<String>,
    pub indexes: Vec<String>,
    pub relationships: HashMap<String, serde_json::Value>,
    pub metadata: HashMap<String, serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ColumnInfo {
    pub name: String,
    pub data_type: DataType,
    pub nullable: bool,
    pub default_value: Option<serde_json::Value>,
    pub constraints: Vec<String>,
    pub description: Option<String>,
    pub format_hint: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum DataType {
    String,
    Number,
    Boolean,
    Date,
    DateTime,
    Array,
    Object,
    Null,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DataQuery {
    pub sql: Option<String>,
    pub path: Option<String>,
    pub filter: Option<String>,
    pub limit: Option<u32>,
    pub offset: Option<u32>,
    pub sort: Option<String>,
    pub parameters: Vec<QueryParameter>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct QueryParameter {
    pub name: String,
    pub param_type: DataType,
    pub default: Option<serde_json::Value>,
    pub description: Option<String>,
}

// === Template Elements ===
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TemplateElement {
    pub id: String,
    pub element_type: ElementType,
    pub position: Position,
    pub size: Size,
    pub z_index: i32,
    pub visible: bool,
    pub content: ElementContent,
    pub style: ElementStyle,
    pub data_binding: Option<DataBinding>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum ElementType {
    Text,
    DataField,
    Rectangle,
    Line,
    Image,
    Barcode,
    Chart,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ElementContent {
    pub text: Option<String>,
    pub expression: Option<String>,
    pub font: Option<FontConfig>,
    pub alignment: Option<AlignmentConfig>,
    pub color: Option<String>,
    pub format: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FontConfig {
    pub family: String,
    pub size: f64,
    pub weight: FontWeight,
    pub style: FontStyle,
}

impl Default for FontConfig {
    fn default() -> Self {
        Self {
            family: "Arial".to_string(),
            size: 12.0,
            weight: FontWeight::Normal,
            style: FontStyle::Normal,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum FontWeight {
    #[serde(rename = "normal")]
    Normal,
    #[serde(rename = "bold")]
    Bold,
    #[serde(rename = "light")]
    Light,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum FontStyle {
    #[serde(rename = "normal")]
    Normal,
    #[serde(rename = "italic")]
    Italic,
    #[serde(rename = "oblique")]
    Oblique,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AlignmentConfig {
    pub horizontal: HorizontalAlignment,
    pub vertical: VerticalAlignment,
}

impl Default for AlignmentConfig {
    fn default() -> Self {
        Self {
            horizontal: HorizontalAlignment::Left,
            vertical: VerticalAlignment::Middle,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum HorizontalAlignment {
    #[serde(rename = "left")]
    Left,
    #[serde(rename = "center")]
    Center,
    #[serde(rename = "right")]
    Right,
    #[serde(rename = "justify")]
    Justify,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum VerticalAlignment {
    #[serde(rename = "top")]
    Top,
    #[serde(rename = "middle")]
    Middle,
    #[serde(rename = "bottom")]
    Bottom,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ElementStyle {
    pub background: Option<BackgroundStyle>,
    pub border: Option<BorderStyle>,
    pub padding: Option<PaddingConfig>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BackgroundStyle {
    pub color: Option<String>,
    pub image: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BorderStyle {
    pub width: f64,
    pub color: String,
    pub style: BorderStyleType,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum BorderStyleType {
    #[serde(rename = "solid")]
    Solid,
    #[serde(rename = "dashed")]
    Dashed,
    #[serde(rename = "dotted")]
    Dotted,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PaddingConfig {
    pub top: f64,
    pub bottom: f64,
    pub left: f64,
    pub right: f64,
}

impl Default for PaddingConfig {
    fn default() -> Self {
        Self {
            top: 0.0,
            bottom: 0.0,
            left: 0.0,
            right: 0.0,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DataBinding {
    pub source_id: String,
    pub field_name: String,
    pub validation: Option<ValidationRules>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ValidationRules {
    pub required: bool,
    pub max_length: Option<usize>,
    pub min_length: Option<usize>,
    pub pattern: Option<String>,
}

// === Parameters and Variables ===
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Parameter {
    pub name: String,
    pub param_type: DataType,
    pub default: Option<serde_json::Value>,
    pub description: Option<String>,
    pub required: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Variable {
    pub name: String,
    pub var_type: DataType,
    pub expression: String,
    pub initial_value: Option<serde_json::Value>,
    pub description: Option<String>,
}

// === Groups ===
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Group {
    pub name: String,
    pub expression: String,
    pub sort_order: SortOrder,
    pub header: Option<BandConfig>,
    pub footer: Option<BandConfig>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum SortOrder {
    #[serde(rename = "asc")]
    Ascending,
    #[serde(rename = "desc")]
    Descending,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BandConfig {
    pub height: f64,
    pub elements: Vec<String>, // Element IDs
}

// === Main Template Structure ===
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JasperTemplate {
    pub metadata: TemplateMetadata,
    pub canvas: Canvas,
    pub data_sources: Vec<DataSource>,
    pub elements: Vec<TemplateElement>,
    pub parameters: Vec<Parameter>,
    pub variables: Vec<Variable>,
    pub groups: Vec<Group>,
}

impl JasperTemplate {
    pub fn new() -> Self {
        Self {
            metadata: TemplateMetadata::default(),
            canvas: Canvas::default(),
            data_sources: Vec::new(),
            elements: Vec::new(),
            parameters: Vec::new(),
            variables: Vec::new(),
            groups: Vec::new(),
        }
    }
    
    pub fn validate(&self) -> Result<()> {
        // Validate canvas
        if self.canvas.width <= 0.0 || self.canvas.height <= 0.0 {
            return Err(AppError::ValidationError {
                message: "Canvas dimensions must be positive".to_string(),
            });
        }
        
        // Validate elements
        for element in &self.elements {
            element.position.validate()?;
            element.size.validate()?;
            
            // Validate data binding references
            if let Some(binding) = &element.data_binding {
                let source_exists = self.data_sources
                    .iter()
                    .any(|ds| ds.id == binding.source_id);
                
                if !source_exists {
                    return Err(AppError::ValidationError {
                        message: format!("Data source '{}' not found for element '{}'", 
                                       binding.source_id, element.id),
                    });
                }
            }
        }
        
        // Validate parameter names are unique
        let mut param_names = std::collections::HashSet::new();
        for param in &self.parameters {
            if !param_names.insert(&param.name) {
                return Err(AppError::ValidationError {
                    message: format!("Duplicate parameter name: {}", param.name),
                });
            }
        }
        
        // Validate variable names are unique
        let mut var_names = std::collections::HashSet::new();
        for var in &self.variables {
            if !var_names.insert(&var.name) {
                return Err(AppError::ValidationError {
                    message: format!("Duplicate variable name: {}", var.name),
                });
            }
        }
        
        Ok(())
    }
    
    pub fn add_element(&mut self, element: TemplateElement) -> Result<()> {
        // Check for duplicate IDs
        if self.elements.iter().any(|e| e.id == element.id) {
            return Err(AppError::ValidationError {
                message: format!("Element with ID '{}' already exists", element.id),
            });
        }
        
        self.elements.push(element);
        self.metadata.last_modified = Utc::now();
        Ok(())
    }
    
    pub fn remove_element(&mut self, element_id: &str) -> Result<bool> {
        let initial_len = self.elements.len();
        self.elements.retain(|e| e.id != element_id);
        
        if self.elements.len() < initial_len {
            self.metadata.last_modified = Utc::now();
            Ok(true)
        } else {
            Ok(false)
        }
    }
    
    pub fn get_element_by_id(&self, element_id: &str) -> Option<&TemplateElement> {
        self.elements.iter().find(|e| e.id == element_id)
    }
    
    pub fn get_element_by_id_mut(&mut self, element_id: &str) -> Option<&mut TemplateElement> {
        self.elements.iter_mut().find(|e| e.id == element_id)
    }
    
    pub fn update_canvas(&mut self, canvas: Canvas) {
        self.canvas = canvas;
        self.metadata.last_modified = Utc::now();
    }
    
    pub fn add_data_source(&mut self, data_source: DataSource) -> Result<()> {
        // Check for duplicate IDs
        if self.data_sources.iter().any(|ds| ds.id == data_source.id) {
            return Err(AppError::ValidationError {
                message: format!("Data source with ID '{}' already exists", data_source.id),
            });
        }
        
        self.data_sources.push(data_source);
        self.metadata.last_modified = Utc::now();
        Ok(())
    }
    
    pub fn get_data_source_by_id(&self, source_id: &str) -> Option<&DataSource> {
        self.data_sources.iter().find(|ds| ds.id == source_id)
    }
}

impl Default for JasperTemplate {
    fn default() -> Self {
        Self::new()
    }
}

// === Conversion utilities ===
impl From<&ReportElement> for TemplateElement {
    fn from(element: &ReportElement) -> Self {
        let (element_type, content) = match &element.content {
            crate::core::element::ElementContent::Text { content, style } => {
                (ElementType::Text, ElementContent {
                    text: Some(content.clone()),
                    expression: None,
                    font: Some(FontConfig {
                        family: style.font_family.clone(),
                        size: style.font_size,
                        weight: match style.font_weight.as_str() {
                            "bold" => FontWeight::Bold,
                            "light" => FontWeight::Light,
                            _ => FontWeight::Normal,
                        },
                        style: FontStyle::Normal,
                    }),
                    alignment: Some(match style.align {
                        crate::core::element::TextAlign::Left => AlignmentConfig {
                            horizontal: HorizontalAlignment::Left,
                            vertical: VerticalAlignment::Middle,
                        },
                        crate::core::element::TextAlign::Center => AlignmentConfig {
                            horizontal: HorizontalAlignment::Center,
                            vertical: VerticalAlignment::Middle,
                        },
                        crate::core::element::TextAlign::Right => AlignmentConfig {
                            horizontal: HorizontalAlignment::Right,
                            vertical: VerticalAlignment::Middle,
                        },
                    }),
                    color: Some(style.color.clone()),
                    format: None,
                })
            }
            crate::core::element::ElementContent::DataField { expression, format, style } => {
                (ElementType::DataField, ElementContent {
                    text: None,
                    expression: Some(expression.clone()),
                    font: Some(FontConfig {
                        family: style.font_family.clone(),
                        size: style.font_size,
                        weight: match style.font_weight.as_str() {
                            "bold" => FontWeight::Bold,
                            "light" => FontWeight::Light,
                            _ => FontWeight::Normal,
                        },
                        style: FontStyle::Normal,
                    }),
                    alignment: Some(match style.align {
                        crate::core::element::TextAlign::Left => AlignmentConfig {
                            horizontal: HorizontalAlignment::Left,
                            vertical: VerticalAlignment::Middle,
                        },
                        crate::core::element::TextAlign::Center => AlignmentConfig {
                            horizontal: HorizontalAlignment::Center,
                            vertical: VerticalAlignment::Middle,
                        },
                        crate::core::element::TextAlign::Right => AlignmentConfig {
                            horizontal: HorizontalAlignment::Right,
                            vertical: VerticalAlignment::Middle,
                        },
                    }),
                    color: Some(style.color.clone()),
                    format: format.clone(),
                })
            }
            crate::core::element::ElementContent::Rectangle { fill_color, .. } => {
                (ElementType::Rectangle, ElementContent {
                    text: None,
                    expression: None,
                    font: None,
                    alignment: None,
                    color: fill_color.clone(),
                    format: None,
                })
            }
            crate::core::element::ElementContent::Line { color, .. } => {
                (ElementType::Line, ElementContent {
                    text: None,
                    expression: None,
                    font: None,
                    alignment: None,
                    color: Some(color.clone()),
                    format: None,
                })
            }
            crate::core::element::ElementContent::Image { src, alt } => {
                (ElementType::Image, ElementContent {
                    text: alt.clone(),
                    expression: Some(src.clone()),
                    font: None,
                    alignment: None,
                    color: None,
                    format: None,
                })
            }
        };
        
        Self {
            id: element.id.to_string(),
            element_type,
            position: element.position.clone(),
            size: element.size.clone(),
            z_index: element.z_index,
            visible: element.visible,
            content,
            style: ElementStyle {
                background: None,
                border: None,
                padding: None,
            },
            data_binding: None,
        }
    }
}